### PR TITLE
refactor(GCS+gRPC): split synthetic self-link helpers

### PR DESCRIPTION
**Pull request by coryan**
Part of the work for #9805

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10806)

### CODE REVIEW

File: google/cloud/storage/internal/grpc_synthetic_self_link.cc [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+std::string SelfLinkPath(Options const& options) {
+  if (!options.has<TargetApiVersionOption>()) return "/storage/v1";
+  return absl::StrCat("/storage/", options.get<TargetApiVersionOption>());
+}
+
+}  // namespace
+
+std::string SyntheticSelfLinkRoot(Options const& options) {
+  return SelfLinkEndpoint(options) + SelfLinkPath(options);
```
Comment (dbolduc) [inline google/cloud/storage/internal/grpc_synthetic_self_link.cc:46]: nit: `absl::StrCat` ?

File: google/cloud/storage/internal/grpc_synthetic_self_link.cc [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+std::string SyntheticSelfLinkBucket(Options const& options,
+                                    std::string const& bucket_name) {
+  return absl::StrCat(SyntheticSelfLinkRoot(options), "/b/", bucket_name);
+}
+
+std::string SyntheticSelfLinkObject(Options const& options,
+                                    std::string const& bucket_name,
+                                    std::string const& object_name) {
+  return absl::StrCat(SyntheticSelfLinkRoot(options), "/b/", bucket_name, "/o/",
+                      UrlEncode(object_name));
```
Comment (dbolduc) [inline google/cloud/storage/internal/grpc_synthetic_self_link.cc:67]: url encoding `object_name` is new (here and below).

If the change is intentional, just ignore this comment.

**Review by dbolduc — APPROVED** [review]

File: google/cloud/storage/internal/grpc_synthetic_self_link.cc (hunk 1 above)
Comment (coryan) [reply google/cloud/storage/internal/grpc_synthetic_self_link.cc]: Fixed.

File: google/cloud/storage/internal/grpc_synthetic_self_link.cc (hunk 2 above)
Comment (coryan) [reply google/cloud/storage/internal/grpc_synthetic_self_link.cc:67]: Sorry, yes.  I fixed a bug along with the refactoring.

**Pull request merged**